tests/lib/prepare-uc.sh

```diff
@@ -0,0 +1,206 @@
+    cp "${artifacts[0]}" "$(get_snap_name)"
+fi
+
+# download snaps required for us to build the image
+download_core24_snaps
+
+# create test user for spread to use
+groupadd --gid 12345 test
+adduser --uid 12345 --gid 12345 --disabled-password --gecos '' test
+
+if getent group systemd-journal >/dev/null; then
+    usermod -G systemd-journal -a test
+    id test | MATCH systemd-journal
+fi
+
+echo 'test ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers
+
+# re-pack snapd snap with special systemd service which runs during run
+# mode to create a user for us to inspect the system state
+
+# TODO actually create a user if needed
+
+snapddir=/tmp/snapd-workdir
+unsquashfs -d $snapddir upstream-snapd.snap
+
+# inject systemd service to setup users and other tweaks for us
+# these are copied from upstream snapd prepare.sh, slightly modified to not 
+# extract snapd spread data from ubuntu-seed as we don't need all that here
+cat > "$snapddir/lib/systemd/system/snapd.spread-tests-run-mode-tweaks.service" <<'EOF'
+[Unit]
+Description=Tweaks to run mode for spread tests
+Before=snapd.service
+Documentation=man:snap(1)
+[Service]
+Type=oneshot
+ExecStart=/usr/lib/snapd/snapd.spread-tests-run-mode-tweaks.sh
+RemainAfterExit=true
+[Install]
+WantedBy=multi-user.target
+EOF
+
+cat > "$snapddir/usr/lib/snapd/snapd.spread-tests-run-mode-tweaks.sh" <<'EOF'
+#!/bin/sh
+set -e
+# Print to kmsg and console
+# $1: string to print
+print_system()
+{
+    printf "%s spread-tests-run-mode-tweaks.sh: %s\n" "$(date -Iseconds --utc)" "$1" |
+        tee -a /dev/ttyS0 /run/mnt/ubuntu-seed/spread-tests-run-mode-tweaks-log.txt || true
+}
+# ensure we don't enable ssh in install mode or spread will get confused
+if ! grep 'snapd_recovery_mode=run' /proc/cmdline; then
+    print_system "not in run mode - script not running"
+    exit 0
+fi
+if [ -e /root/spread-setup-done ]; then
+    print_system "already ran, not running again"
+    exit 0
+fi
+print_system "in run mode, not run yet, extracting overlay data"
+# extract data from previous stage
+if [ -f /run/mnt/ubuntu-seed/run-mode-overlay-data.tar.gz ]; then
+   (cd / && tar xf /run/mnt/ubuntu-seed/run-mode-overlay-data.tar.gz)
+   cp -r /root/test-var/lib/extrausers /var/lib
+   # user db - it's complicated
+   for f in group gshadow passwd shadow; do
+       # now bind mount read-only those passwd files on boot
+       if [ ! -f /root/test-etc/$f ]; then
+          continue
+       fi
+       cat >/etc/systemd/system/etc-"$f".mount <<EOF2
+[Unit]
+Description=Mount root/test-etc/$f over system etc/$f
+Before=ssh.service
+[Mount]
+What=/root/test-etc/$f
+Where=/etc/$f
+Type=none
+Options=bind,ro
+[Install]
+WantedBy=multi-user.target
+EOF2
+        systemctl enable etc-"$f".mount
+        systemctl start etc-"$f".mount
+    done
+fi
+# TODO: do we need this for our nested VM? We don't login as root to the nested
+#       VM...
+sed -i 's/\#\?\(PermitRootLogin\|PasswordAuthentication\)\>.*/\1 yes/' /etc/ssh/sshd_config
```

Comment:
Nitpick, you can use `s,foo,bar,` to avoid some of the escape, perhaps?

Reply:
I would love if we had vsock support for looking into test VMs that don't have network and have locked down ssh/serial/what-not.